runtime/master/src/main/java/edu/snu/nemo/runtime/master/BlockManagerMaster.java

```diff
@@ -43,31 +37,32 @@
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
-import java.util.stream.IntStream;
+import java.util.stream.Collectors;
 
 import org.apache.reef.annotations.audience.DriverSide;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import static edu.snu.nemo.runtime.common.state.BlockState.State.IN_PROGRESS;
-import static edu.snu.nemo.runtime.common.state.BlockState.State.NOT_AVAILABLE;
-
 /**
  * Master-side block manager.
  */
 @ThreadSafe
 @DriverSide
 public final class BlockManagerMaster {
   private static final Logger LOG = LoggerFactory.getLogger(BlockManagerMaster.class.getName());
-  private final Map<String, BlockMetadata> blockIdToMetadata;
-  private final Map<String, Set<String>> producerTaskIdToBlockIds;
+
+  private final Map<String, Set<String>> producerTaskIdToBlockIds; // a task can have multiple out-edges
+  private final Map<String, Set<BlockMetadata>> blockIdWildcardToMetadataSet; // a metadata = a task attempt output
```

Comment:
Could you explain more on the usage of `blockIdWildcardToMetadataSet`, i.e. why it is needed and what exactly 'a task attempt output' means? For example, does 'block wildcard' correspond to output of certain task attempt or cloned task?